tokio/src/util/wake_list.rs

```diff
@@ -0,0 +1,47 @@
+use core::mem::MaybeUninit;
+use std::task::Waker;
+
+const NUM_WAKERS: usize = 32;
+
+pub(crate) struct WakeList {
+    inner: [MaybeUninit<Waker>; NUM_WAKERS],
+    curr: usize,
+}
+
+impl WakeList {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: unsafe { MaybeUninit::uninit().assume_init() },
+            curr: 0,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn can_push(&self) -> bool {
+        self.curr < NUM_WAKERS
+    }
+
+    pub(crate) fn push(&mut self, val: Waker) {
+        debug_assert!(self.can_push());
+
+        self.inner[self.curr] = MaybeUninit::new(val);
+        self.curr += 1;
+    }
+
+    pub(crate) fn wake_all(&mut self) {
+        assert!(self.curr <= NUM_WAKERS);
```

Comment:
We may want to make this a `debug_assert!`, as the indexing on line 35 will panic if `self.curr` is out of bounds. The assertion is mainly just giving us a nicer panic message at the cost of additional test and jump instructions.

Reply:
Wouldn't it affect the optimization of the release build?

Reply:
Nevermind, as per https://github.com/tokio-rs/tokio/pull/4055#discussion_r694218623, the presence of the assertion actually influences optimizer behavior --- let's not remove it.